File: examples/24-monolith-to-microservices/python/router/main.py
import os

import httpx
from fastapi import FastAPI, Request
from starlette.responses import Response

app = FastAPI()

MONOLITH_URL = os.environ.get("MONOLITH_URL", "http://monolith:8080")
NEW_SERVICE_URL = os.environ.get("NEW_SERVICE_URL", "http://new-service:8080")
client = httpx.AsyncClient(timeout=5.0)

routes: dict = {
    "tenant_routes": {"acme": "new-service"},
    "default": "monolith",
}


def _resolve_upstream(tenant: str) -> str:
    target = routes["tenant_routes"].get(tenant, routes["default"])
    if target == "new-service":
        return NEW_SERVICE_URL
    return MONOLITH_URL


@app.api_route("/orders", methods=["POST"])
async def route_post(request: Request):
    body = await request.body()
    import json

    payload = json.loads(body) if body else {}
    tenant = payload.get("tenant", "")
    upstream = _resolve_upstream(tenant)
    r = await client.request(
        request.method,
        f"{upstream}/orders",
        content=body,
        headers={"content-type": "application/json"},
    )
    return Response(content=r.content, status_code=r.status_code)


@app.get("/orders/{order_id}")
async def route_get(order_id: str, tenant: str = ""):
    upstream = _resolve_upstream(tenant)
    r = await client.get(f"{upstream}/orders/{order_id}")
    return Response(content=r.content, status_code=r.status_code)


@app.get("/rules")
async def get_rules():
    return routes


@app.put("/rules")
async def update_rules(new_rules: dict):
    routes.update(new_rules)
    return routes


@app.get("/healthz")
async def healthz():
    return {"status": "ok"}

File: examples/24-monolith-to-microservices/cpp/router/src/main.cpp
#include <drogon/drogon.h>
#include <mutex>
#include <unordered_map>
#include <iostream>

using namespace drogon;

static std::mutex g_mu;
static std::unordered_map<std::string, std::string> g_tenant_routes;
static std::string g_default_service;
static std::unordered_map<std::string, std::string> g_service_urls;

static void forward_request(const std::string& upstream, HttpMethod method,
                            const std::string& path, const std::string& body,
                            std::function<void(const HttpResponsePtr&)> cb) {
    auto client = HttpClient::newHttpClient(upstream);
    auto req = HttpRequest::newHttpRequest();
    req->setMethod(method);
    req->setPath(path);
    if (!body.empty()) {
        req->setContentTypeCode(CT_APPLICATION_JSON);
        req->setBody(body);
    }
    client->sendRequest(req,
        [cb = std::move(cb)](ReqResult result, const HttpResponsePtr& resp) {
            if (result == ReqResult::Ok) {
                auto r = HttpResponse::newHttpResponse();
                r->setStatusCode(resp->statusCode());
                r->setContentTypeCode(CT_APPLICATION_JSON);
                r->setBody(std::string(resp->body()));
                cb(r);
            } else {
                auto r = HttpResponse::newHttpResponse();
                r->setStatusCode(k502BadGateway);
                cb(r);
            }
        });
}

static std::string resolve_upstream(const std::string& tenant) {
    std::lock_guard<std::mutex> lock(g_mu);
    auto it = g_tenant_routes.find(tenant);
    std::string svc = (it != g_tenant_routes.end()) ? it->second : g_default_service;
    auto uit = g_service_urls.find(svc);
    if (uit != g_service_urls.end()) return uit->second;
    return g_service_urls[g_default_service];
}

int main() {
    auto env_m = std::getenv("MONOLITH_URL");
    auto env_n = std::getenv("NEW_SERVICE_URL");
    g_service_urls["monolith"] = env_m ? env_m : "http://monolith:8080";
    g_service_urls["new-service"] = env_n ? env_n : "http://new-service:8080";

    g_tenant_routes["acme"] = "new-service";
    g_default_service = "monolith";

    app().setLogLevel(trantor::Logger::kInfo);
    app().addListener("0.0.0.0", 8080);

    app().registerHandler("/healthz",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& cb) {
            Json::Value r;
            r["status"] = "ok";
            cb(HttpResponse::newHttpJsonResponse(r));
        },
        {Get});

    app().registerHandler("/orders",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& cb) {
            auto body = req->getJsonObject();
            std::string tenant;
            if (body && body->isMember("tenant")) {
                tenant = (*body)["tenant"].asString();
            }
            std::string upstream = resolve_upstream(tenant);
            std::cerr << "routing tenant=" << tenant << " to " << upstream << std::endl;
            forward_request(upstream, Post, "/orders", std::string(req->body()), std::move(cb));
        },
        {Post});

    app().registerHandler("/orders/{order_id}",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& cb,
           const std::string& order_id) {
            std::string upstream = resolve_upstream("");
            forward_request(upstream, Get, "/orders/" + order_id, "", std::move(cb));
        },
        {Get});

    app().registerHandler("/rules",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& cb) {
            std::lock_guard<std::mutex> lock(g_mu);
            Json::Value r;
            Json::Value tr(Json::objectValue);
            for (auto& [k, v] : g_tenant_routes) tr[k] = v;
            r["tenant_routes"] = tr;
            r["default"] = g_default_service;
            cb(HttpResponse::newHttpJsonResponse(r));
        },
        {Get});

    app().registerHandler("/rules",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& cb) {
            auto body = req->getJsonObject();
            if (!body) {
                auto resp = HttpResponse::newHttpResponse();
                resp->setStatusCode(k400BadRequest);
                cb(resp);
                return;
            }
            {
                std::lock_guard<std::mutex> lock(g_mu);
                if (body->isMember("tenant_routes")) {
                    g_tenant_routes.clear();
                    auto& tr = (*body)["tenant_routes"];
                    for (auto it = tr.begin(); it != tr.end(); ++it) {
                        g_tenant_routes[it.key().asString()] = it->asString();
                    }
                }
                if (body->isMember("default")) {
                    g_default_service = (*body)["default"].asString();
                }
            }
            std::cerr << "rules updated" << std::endl;
            Json::Value r;
            r["status"] = "updated";
            cb(HttpResponse::newHttpJsonResponse(r));
        },
        {Put});

    app().run();
    return 0;
}
